mod_proxy_http: handle Upgrade request and Switching Protocol

ylavic opened the pull request:
If the request Upgrade header matches the worker parameter "upgrade" and
the backend switches the protocol, do the tunneling in mod_proxy_http.
This allows to keep the protocol to HTTP until the backend really
switches the protocol, to apply usual output filters.

ylavic committed:
mod_proxy_http: handle Upgrade request and 101 (Switching Protocol) response.

If the request Upgrade header matches the worker parameter "upgrade" and
the backend switches the protocol, do the tunneling in mod_proxy_http.
This allows to keep the protocol to HTTP until the backend really
switches the protocol, to apply usual output filters.

rpluem commented on modules/proxy/mod_proxy_http.c:
```diff
@@ -1615,6 +1649,62 @@ int ap_proxy_http_process_response(proxy_http_req_t *req)
+            else {
+                tunnel->timeout = client_timeout;
+            }
+
```
Why do we need to align the timeouts and and have a procedure different from mod_proxy_wstunnel? This can be confusing.

ylavic replied:
In mod_proxy_wstunnel there is also a single (idle_)timeout for the tunneling part, this is how the ap_proxy_tunnel loop works (I don't it makes sense to have two separate timeouts for client and backend once it's not HTTP anymore).

Whether we should use the min, max, always use the backend's, or have a dedicated idle timeout is an open question, though. I'd say the backend's, maybe?

rpluem replied:
I saw that a single timeout is used in `ap_proxy_tunnel` and this is not the point I wanted to tackle. I was more worried that people get confused timeout wise with the setting done by `ProxyWebsocketIdleTimeout` which is `mod_proxy_wstunnel` related.  But this is not backported yet and probably before it is backported / 2.5[6] is released we have all that in `mod_proxy_http` such there is no need for having `ProxyWebsocketIdleTimeout`  ever in a stable release in its current form.
Once we have all that in `mod_proxy_http` we can think of how to handle it with the same parameters (approach) we currently use for timeouts. I think we already use two timeouts in `mod_proxy_wstunnel`in trunk. One for idle timeouts and one (or two different ones for client and server) for blocking timeouts.

ylavic replied:
The plan would indeed be to deprecate proxy_wstunnel and have the upgrades handled by proxy_http, so `ProxyWebsocketIdleTimeout` would become ProxyAsyncTimeout and for the HTTP protocol we'd keep the usual `Timeout` (client side) and `ProxyTimeout` / worker `timeout=` on the origin side (which are the ones used by proxy_wstunnel too until the upgrade).

I'm on the async side of things now for this to possible ;)
Though as you said this is trunk only today in proxy_wstunnel.

rpluem commented on modules/proxy/mod_proxy_http.c:
```diff
@@ -1615,6 +1649,62 @@ int ap_proxy_http_process_response(proxy_http_req_t *req)
+
+            /* Let proxy tunnel forward everything */
+            status = ap_proxy_tunnel_run(tunnel);
+            if (ap_is_HTTP_ERROR(status)) {
```
The procedure here looks different from the one in mod_proxy_ws_tunnel:

1. We don't support async MPM's
2. We do not remove the reqtimeout filter

ylavic replied:
1. Async is coming next (probably a separate PR).
This would not be only for the tunneling part, we can go async for HTTP too when/if waiting for client (body) or backend data.
I need to find the right interface in proxy_util (based on wstunnel pump mechanism) for any proxy module to be able to use it.

2. An oversight, will do it now.

ylavic replied:
Point 2. addressed in 5228c13.

ylavic committed:
mod_proxy_http: disable reqtimeout for Upgrade(d) connections.

covener commented:
Curious, would you keep the current async method or move the work to a filter?

ylavic commented:
> Curious, would you keep the current async method or move the work to a filter?

Still the MPM even callback interface like wstunnel, the filters' should_yield() and co would help determine whether we need to go async only.

ylavic committed:
mod_proxy_http: forward back the Upgrade header matching the configuration.

When configured to forward Upgrade mechanism, we want the backend to be
able to announce its Upgrade protocol to the client (e.g. with 426
Upgrade Required response) and thus forward back the Upgrade header that
matches the one(s) configured in the worker upgrade= parameter.

include/ap_mmn.h:
modules/proxy/mod_proxy.h:
modules/proxy/proxy_util.c:
    Add ap_proxy_worker_can_upgrade() helper to determine whether a
    proxy worker is configured to forward an Upgrade protocol.
    Bump MMN minor.

modules/proxy/mod_proxy.c:
    Handle worker parameter upgrade=ANY as upgrade=* (should the "any"
    protocol scheme be something some day..).

modules/proxy/mod_proxy_wstunnel.c:
    Use ap_proxy_worker_can_upgrade() to match the Upgrade header in
    proxy_wstunnel_handler().
    Axe handling of upgrade=NONE, it makes no sense to Upgrade a
    connection if the client did not ask for it, nor to configure
    mod_proxy_wstunnel to use a worker with upgrade=NONE by the way.

modules/proxy/mod_proxy_http.c:
    Use ap_proxy_worker_can_upgrade() to match the Upgrade header in
    proxy_http_handler() and ap_proxy_http_process_response().
    Forward back the Upgrade header in ap_proxy_http_process_response()
    if it matches the configured one(s). That is for 101 Switching
    Protocol obviously but also any other status code which is not
    overidden, at the backend wish.

ylavic committed:
Update APLOGNOs.

minfrin commented:
Backported to v2.4.47.
